RavenDB-10724 add raven client for google cloud storage backup & test.

Or-Geva opened the pull request.

Commit:
RavenDB-10724 add raven client for google cloud storage backup & test.
Split AWS & Azure tests into two separate cs files

grisha-kotler commented on src/Raven.Client/Documents/Operations/Backups/PeriodicBackupConfiguration.cs:
```diff
@@ -121,6 +125,8 @@ public List<string> GetDestinations()
                 backupDestinations.Add("S3");
             if (GlacierSettings != null && GlacierSettings.Disabled == false)
                 backupDestinations.Add("Glacier");
+            if (FtpSettings != null && GoogleCloudStorageSettings.Disabled == false)
```
```suggestion
            if (GoogleCloudStorageSettings != null && GoogleCloudStorageSettings.Disabled == false)
```

Or-Geva replied:
Fixed

grisha-kotler commented on src/Raven.Server/Documents/PeriodicBackup/GoogleCloudStorage/RavenGoogleCloudStorageClient.cs:
```diff
@@ -0,0 +1,160 @@
+                    _progress.UploadProgress.UploadedInBytes = p.BytesSent;
+                    _progress.OnUploadProgress();
+
+                    Console.WriteLine($"total bytes {stream.Length}, uploaded {p.BytesSent}, status: {p.Status}");
```
Can be removed

Or-Geva replied:
Fixed

grisha-kotler commented on test/Tests.Infrastructure/GoogleCloudStorageFact.cs:
```diff
@@ -0,0 +1,38 @@
+
+        static GoogleCloudStorageFact()
+        {
+            Environment.SetEnvironmentVariable("GOOGLE_CLOUD_STORAGE_CREDENTIAL","{\n  \"type\": \"service_account\",\n  \"project_id\": \"prefab-phoenix-237012\",\n  \"private_key_id\": \"e6b23b33d06dd37025054cbe8f18e57f081c6d72\",\n  \"private_key\": \"-----BEGIN PRIVATE KEY-----\\nMIIEvQIBADANBgkqhkiG9w0BAQEFAASCBKcwggSjAgEAAoIBAQCgVuKV59aV9UGB\\nx/EHF0rXj4FYK9vESvdjCT5tK6cvcotxWS7tmydvucLJt7H ...
```
Can be removed

Or-Geva replied:
Fixed

grisha-kotler commented on src/Raven.Studio/typescript/models/database/tasks/periodicBackup/googleCloudStorageSettings.ts:
```diff
@@ -0,0 +1,108 @@
+            validation: [
+                {
+                    validator: (GoogleCredentialsJson: string) => this.validate(() =>
+                    GoogleCredentialsJson.includes("\"type\"")),
```
Is "Type" is also accepted?

Or-Geva replied:
Yes

Commit:
RavenDB-10724 fix PR comments

ml054 commented on src/Raven.Studio/typescript/models/database/tasks/periodicBackup/googleCloudStorageSettings.ts:
```diff
@@ -0,0 +1,108 @@
+class googleCloudStorageSettings extends backupSettings {
+    bucket = ko.observable<string>();
+    remoteFolderName = ko.observable<string>();
+    GoogleCredentialsJson = ko.observable<string>();
```
use lowercase letter in variable name

Or-Geva replied:
Fixed

ml054 commented on src/Raven.Studio/wwwroot/App/views/database/tasks/editPeriodicBackupTask.html:
```diff
@@ -346,6 +346,67 @@ <h3>Destination</h3>
+                    <div class="text-right">
+                        <button data-bind="click: _.partial($root.testCredentials, $data), disable: isTestingCredentials, css: { 'btn-spinner': isTestingCredentials }" class="btn btn-info">
+                            <i class="icon-rocket"></i>
+                            &nbsp;&nbsp;<span>Test credentials</span>
```
remove those nbsp

Or-Geva replied:
Fixed

ayende commented on src/Raven.Server/Documents/PeriodicBackup/GoogleCloudStorage/RavenGoogleCloudStorageClient.cs:
```diff
@@ -0,0 +1,158 @@
+            catch (Google.GoogleApiException e)
+                when (e.Error.Code == 403)
+            {
+                throw new InvalidOperationException($"Google credentials json does not have access to project {_projectId ?? "N/A"}");
```
You need to include the actual exception in _all_ the exceptions you rethrow

Or-Geva replied:
Fixed

Commit:
RavenDB-10724 fix pr comments & regex

Commit:
RavenDB-10724 fix pr comments - include the actual exception in all the exceptions.

arekpalinski commented:
Let's hold on with merging this until @ppekrol will review it as well.

ppekrol commented on src/Raven.Client/Documents/Operations/Backups/BackupProgress.cs:
```diff
@@ -11,6 +11,8 @@ public class BackupProgress : SmugglerResult.SmugglerProgress
 
         public UploadToAzure AzureBackup => ((BackupResult)_result).AzureBackup;
 
+        public UploadToGoogleCloudStorage GoogleCloudStorageBackup => ((BackupResult)_result).GoogleCloudStorageBackup;
```
Rename:
- type to `UploadToGoogleCloud` 
- property to `GoogleCloudBackup`

Or-Geva replied:
Done.

ppekrol commented on src/Raven.Client/Documents/Operations/Backups/BackupSettings.cs:
```diff
@@ -241,6 +241,50 @@ public override DynamicJsonValue ToJson()
+            return djv;
+        }
+    }
+    public class GoogleCloudStorageSettings : BackupSettings
```
`GoogleCloudSettings`

Or-Geva replied:
Done.

ppekrol commented on src/Raven.Client/Documents/Operations/Backups/PeriodicBackupConfiguration.cs:
```diff
@@ -121,6 +125,8 @@ public List<string> GetDestinations()
             if (GlacierSettings != null && GlacierSettings.Disabled == false)
                 backupDestinations.Add("Glacier");
+            if (GoogleCloudStorageSettings != null && GoogleCloudStorageSettings.Disabled == false)
+                backupDestinations.Add("Google Cloud Storage");
```
just `Google Cloud`

Or-Geva replied:
Done.

ppekrol commented on src/Raven.Server/Config/Categories/BackupConfiguration.cs:
```diff
@@ -17,7 +17,7 @@ public class BackupConfiguration : ConfigurationCategory
         public PathSetting LocalRootPath { get; set; }
         
-        [Description("Semicolon seperated list of allowed backup destinations. If not specified, all destinations are allowed. Possible values: None, Local, Azure, AmazonGlacier, AmazonS3, FTP. Example list: \"Local;AmazonGlacier;AmazonS3\".")]
+        [Description("Semicolon seperated list of allowed backup destinations. If not specified, all destinations are allowed. Possible values: None, Local, Azure, AmazonGlacier, AmazonS3,GoogleCloudStorage, FTP. Example list: \"Local;AmazonGlacier;AmazonS3\".")]
```
missing space :)

Or-Geva replied:
Done.

ppekrol commented on src/Raven.Server/Config/Categories/BackupConfiguration.cs:
```diff
@@ -41,7 +41,7 @@ public override void Initialize(IConfigurationRoot settings, IConfigurationRoot
         private readonly HashSet<string> _allDestinations = new HashSet<string>
         {
-            "None", "Local", "Azure", "AmazonGlacier", "AmazonS3", "FTP"
+            "None", "Local", "Azure", "AmazonGlacier", "AmazonS3", "FTP", "GoogleCloudStorage"
```
`GoogleCloud`

Or-Geva replied:
Done.

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/GoogleCloudStorage/RavenGoogleCloudStorageClient.cs:
```diff
@@ -0,0 +1,158 @@
+                throw new ArgumentException("Wrong format for account key.", e);
+            }
+
+            if (JObject.Parse(jsonCredential).TryGetValue(ProjectIdPropertyName, StringComparison.InvariantCultureIgnoreCase, out var value))
```
Let's split this to two operations. Also use OridinalIgnoreCase comparer

Or-Geva replied:
Done.

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/GoogleCloudStorage/RavenGoogleCloudStorageClient.cs:
```diff
@@ -0,0 +1,158 @@
+        public async Task<Object> UploadObjectAsync(string fileName, Stream stream, Dictionary<string, string> metadata = null)
+        {
+
+            return await _client.UploadObjectAsync(
```
why awaiting?

Or-Geva replied:
To return the object & metadata

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/GoogleCloudStorage/RavenGoogleCloudStorageClient.cs:
```diff
@@ -0,0 +1,158 @@
+
+        public async Task DownloadObjectAsync(string fileName, Stream stream)
+        {
+            await _client.DownloadObjectAsync(
```
why awaiting?

Or-Geva replied:
So the stream will finish downloading the object.

grisha-kotler replied:
Don't you need to wait for the object to finish downloading?

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/GoogleCloudStorage/RavenGoogleCloudStorageClient.cs:
```diff
@@ -0,0 +1,158 @@
+
+        public async Task<Object> GetObjectAsync(string fileName)
+        {
+            return await _client.GetObjectAsync(
```
why awaiting?

Or-Geva replied:
To return the object & metadata

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/GoogleCloudStorage/RavenGoogleCloudStorageClient.cs:
```diff
@@ -0,0 +1,158 @@
+
+        public async Task DeleteObjectAsync(string fileName)
+        {
+            await _client.DeleteObjectAsync(
```
why awaiting?

Postyy replied:
Removed & change to void.

Or-Geva replied:
We need to wait before starting a new test

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/GoogleCloudStorage/RavenGoogleCloudStorageClient.cs:
```diff
@@ -0,0 +1,158 @@
+            catch (Google.GoogleApiException e)
+                when (e.Error.Code == 404)
+            {
+                throw new InvalidOperationException($"Bucket {_bucketName} not found!", e);
```
Remove the `!`

Or-Geva replied:
Done.

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/PeriodicBackupTestConnectionType.cs:
```diff
@@ -8,6 +8,7 @@ public enum PeriodicBackupTestConnectionType
         S3,
         Glacier,
         Azure,
+        GoogleCloudStorage,
```
just `GoogleCloud`

Or-Geva replied:
Done.

ppekrol commented:
Also 60 tests are failing in this PR. Please check the reason.

Commit:
RavenDB-10724 Refactor GoogleCloudStorage -> GoogleCloud | add delete test | improve listObject

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/GoogleCloud/RavenGoogleCloudClient.cs:
```diff
@@ -0,0 +1,161 @@
+        public async Task<Object> UploadObjectAsync(string fileName, Stream stream, Dictionary<string, string> metadata = null)
+        {
+
+            return await _client.UploadObjectAsync(
```
you can remove `async/await` and just return the task from the client

Or-Geva replied:
Done.

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/GoogleCloud/RavenGoogleCloudClient.cs:
```diff
@@ -0,0 +1,161 @@
+
+        public async Task DownloadObjectAsync(string fileName, Stream stream)
+        {
+            await _client.DownloadObjectAsync(
```
you can remove `async/await` and just return the task from the client

Or-Geva replied:
Done.

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/GoogleCloud/RavenGoogleCloudClient.cs:
```diff
@@ -0,0 +1,161 @@
+
+        public async Task<Object> GetObjectAsync(string fileName)
+        {
+            return await _client.GetObjectAsync(
```
you can remove `async/await` and just return the task from the client

Or-Geva replied:
Done.

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/GoogleCloud/RavenGoogleCloudClient.cs:
```diff
@@ -0,0 +1,161 @@
+
+        public async Task DeleteObjectAsync(string fileName)
+        {
+            await _client.DeleteObjectAsync(
```
you can remove `async/await` and just return the task from the client

Or-Geva replied:
Done.

Commit:
RavenDB-10724 fix PR comment | remove async/await

ppekrol committed:
RavenDB-10724
- added Microsoft.Extensions.Configuration.UserSecrets so project will build correctly on 3.0 preview SDKs
- added Google.Api.Gax.Rest so the dependency of Google.Cloud.Storage.V1 will be bumped
